### PR TITLE
Modular saltworks

### PR DESCRIPTION
**Pull request by Maleclypse**
#### Summary
Summary: Content "Faction Camp Modular Saltworks"

#### Purpose of change
Creating an additional modular camp expansion.  Modeled after Curstwist's PRs in the same topic.  

#### Describe the solution
Creating additional json files similar the livestock and storage campsite PR's.  

#### Describe alternatives you've considered
Continuing to just make single json edits. 

#### Additional context
"WIP" I've managed to get the game to load without load stopping errors but I can't get the expand base mission to complete successfully and create the modular campsite Saltworks. Any suggestions very welcome.  
Once I can get one version to create I plan to create other building material versions and work on the recipe mission list. 

I'm pretty ready to close this out but I think I'm going to need some help to finish it in a reasonable timeline.

### CODE REVIEW

**Comment by Maleclypse** [comment]
Error recipe group
failed to add saltworks

File: data/json/recipes/basecamps/recipe_modular_saltworks_log.json [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    "autolearn": false,
+    "never_learn": true,
+    "construction_blueprint": "fbms_brewery2_log_southwest",
+    "blueprint_name": "northwest stable",
+    "blueprint_requires": [ { "id": "fbms_southwest" } ],
+    "blueprint_provides": [ { "id": "fbms_southwest" } ],
+    "blueprint_excludes": [ { "id": "fbms_souththwest" } ],
+    "blueprint_autocalc": true
+  }
+]
```
Comment (ZhilkinSerg) [inline data/json/recipes/basecamps/recipe_modular_saltworks_log.json]:
```suggestion
]

```

File: data/json/mapgen/basecamps/modular_saltworks_log.json [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      "palettes": [ "fbms_log_palette" ]
+    }
+  },
+  {
+    "type": "mapgen",
+    "update_mapgen_id": "fbms_brewery2_log_west",
+    "method": "json",
+    "object": { "place_nested": [ { "chunks": [ "fbml_stable2_log" ], "x": 3, "y": 14 } ] }
+  }
+]
```
Comment (ZhilkinSerg) [inline data/json/mapgen/basecamps/modular_saltworks_log.json]:
```suggestion
]

```

**Comment by Maleclypse** [comment]
@curstwist I hate to bother you, but if you have a second to look over this at any time and if you see where I'm going wrong I'd appreciate the help.  If no time, no worries I'll muddle through :)

**Comment by curstwist** [comment]
Looks like you are missing the survey recipe step.

**Comment by Maleclypse** [comment]
> Looks like you are missing the survey recipe step.
> 
> And its mapgen partner. Look at the 1st entries in a _wood recipe &mapgen file. I dont repeat it in other palettes
> 
> ```
> {
>         "type": "recipe",
>         "result": "faction_base_workshop_0",
>         "description": "Survey land for a workshop area.",
>         "category": "CC_BUILDING",
>         "subcategory": "CSC_BUILDING_BASES",
>         "skill_used": "fabrication",
>         "autolearn": false,
>         "never_learn": true,
>         "construction_blueprint": "faction_base_workshop_0",
>         "blueprint_name": "workshop survey",
>         "time": "180 m",
>         "blueprint_requires": [ { "id": "not_an_upgrade" } ],
>         "blueprint_provides": [ { "id": "fbmw_0" } ]
>       },
> ```
> 
> and
> 
> ```
> {
>     "type": "mapgen",
>     "method": "json",
>     "nested_mapgen_id": "fbmw_0",
>     "object": {
>       "mapgensize": [ 6, 6 ],
>       "rows": [
>         "      ",
>         "      ",
>         "      ",
>         "      ",
>         "      ",
>         "      "
>       ],
>       "palettes": [ "fbmw_wood_palette" ]
>     }
>   },
>   {
>     "type": "mapgen",
>     "update_mapgen_id": "faction_base_workshop_0",
>     "method": "json",
>     "object": { "place_nested": [ { "chunks": [ "fbml_0" ], "x": 15, "y": 3 } ] }
>   },
> ```
> 
> that little mapgen really doesn't do anything to the map, but I got errors if I left it out, the survey wants an update_mapgen_id.

Thanks! I'll work on that this week.

**Comment by Maleclypse** [comment]
I'm having a JSON validation error " make: *** [data/json/recipes/basecamps/recipe_groups.json] Error 1
Makefile:1034: recipe for target 'data/json/recipes/basecamps/recipe_groups.json' failed" but I'm having difficulty figuring out any differences between my version and the baseline version that would be causing this to fail?

File: data/json/recipes/basecamps/recipe_modular_saltworks_stone.json [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        "      ",
+        "      ",
+        "      "
+      ],
+      "palettes": [ "fbms_rock_palette" ]
+    }
+  },
+  {
+    "type": "mapgen",
+    "update_mapgen_id": "faction_base_workshop_0",
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks_stone.json]: `faction_base_workshop_0` should probably be `faction_base_saltworks_0`

File: data/json/recipes/basecamps/recipe_modular_saltworks_wad.json [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        "      ",
+        "      ",
+        "      "
+      ],
+      "palettes": [ "fbms_wad_palette" ]
+    }
+  },
+  {
+    "type": "mapgen",
+    "update_mapgen_id": "faction_base_workshop_0",
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks_wad.json]: `faction_base_saltworks_0` again

File: data/json/recipes/basecamps/recipe_modular_saltworks_wood.json [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        "      ",
+        "      ",
+        "      "
+      ],
+      "palettes": [ "fbms_wood_palette" ]
+    }
+  },
+  {
+    "type": "mapgen",
+    "update_mapgen_id": "faction_base_workshop_0",
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks_wood.json]: correct to: `faction_base_saltworks_0`

File: data/json/mapgen/basecamps/modular_saltworks_wad.json [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        "      ",
+        "      ",
+        "      "
+      ],
+      "palettes": [ "fbms_wad_palette" ]
+    }
+  },
+  {
+    "type": "mapgen",
+    "update_mapgen_id": "faction_base_workshop_0",
```
Comment (curstwist) [inline data/json/mapgen/basecamps/modular_saltworks_wad.json]: `faction_base_saltworks_0`

**Comment by curstwist** [comment]
> 
> 
> I'm having a JSON validation error " make: *** [data/json/recipes/basecamps/recipe_groups.json] Error 1
> Makefile:1034: recipe for target 'data/json/recipes/basecamps/recipe_groups.json' failed" but I'm having difficulty figuring out any differences between my version and the baseline version that would be causing this to fail?

you have some references to the workshop that could be the problem.  Also, you only need the survey recipe and its mapgen once, no need to repeat it in every palette.

**Comment by Maleclypse** [comment]
All checks clear and testing works. Do I need to close this PR and open a nonWIP version?

**Comment by anothersimulacrum** [comment]
Nope, just remove WIP in the title.

**Comment by Night-Pryanik** [comment]
Some screenshots?

**Comment by Maleclypse** [comment]
> Some screenshots?


I edited the descriptions.  I also noticed that the brewery upgrade is failing and not calculating the components needed to build so if anyone has any suggestions on how to fix that I'm open to them.  I'll turn this back into WIP until  I get that resolved. When I tested it the first time I just got them all built to first level.  

Saltworks first level
failed
recipe failure

**Comment by Maleclypse** [comment]
@curstwist were you having issues with upgrades to existing buildings and that's why you stopped making basecamp upgrades until it was resolved?  I can't remember.

**Comment by curstwist** [comment]
> 
> 
> @curstwist were you having issues with upgrades to existing buildings and that's why you stopped making basecamp upgrades until it was resolved? I can't remember.

I stopped until the rotation issue is fixed (there's a PR up that should fix it) and using the expansion upgrade craft recipies in the "core" building, which isn't working.  Since this is a proper expansion, it shouldn't be either of my issues.  I'll take a look at it.

File: data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_wad.json [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    "description": "We could use fill out the space with extra kegs and storage racks.",
+    "category": "CC_BUILDING",
+    "subcategory": "CSC_BUILDING_BASES",
+    "autolearn": false,
+    "never_learn": true,
+    "construction_blueprint": "fbmsw_brewery2_wad_southwest",
+    "blueprint_name": "southwest brewery2",
+    "blueprint_requires": [ { "id": "fbmsw_southwest" } ],
+    "blueprint_provides": [ { "id": "fbmsw_southwest" } ],
+    "blueprint_excludes": [ { "id": "fbmsw_souththwest" } ],
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_wad.json]: typo in blueprint excludes southtwest

File: data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_metal.json [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    "description": "We could use fill out the space with extra kegs and storage racks.",
+    "category": "CC_BUILDING",
+    "subcategory": "CSC_BUILDING_BASES",
+    "autolearn": false,
+    "never_learn": true,
+    "construction_blueprint": "fbmsw_brewery2_metal_southwest",
+    "blueprint_name": "southwest brewery2",
+    "blueprint_requires": [ { "id": "fbmsw_southwest" } ],
+    "blueprint_provides": [ { "id": "fbmsw_southwest" } ],
+    "blueprint_excludes": [ { "id": "fbmsw_souththwest" } ],
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_metal.json]: "blueprint_excludes": [ { "id": "fbmsw_southwest" } ],

File: data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_stone.json [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    "description": "We could use fill out the space with extra kegs and storage racks.",
+    "category": "CC_BUILDING",
+    "subcategory": "CSC_BUILDING_BASES",
+    "autolearn": false,
+    "never_learn": true,
+    "construction_blueprint": "fbmsw_brewery2_rock_southwest",
+    "blueprint_name": "southwest brewery2",
+    "blueprint_requires": [ { "id": "fbmsw_southwest" } ],
+    "blueprint_provides": [ { "id": "fbmsw_southwest" } ],
+    "blueprint_excludes": [ { "id": "fbmsw_souththwest" } ],
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_stone.json]: "blueprint_excludes": [ { "id": "fbmsw_souththwest" } ],

File: data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_wood.json [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    "description": "We could use fill out the space with extra kegs and storage racks.",
+    "category": "CC_BUILDING",
+    "subcategory": "CSC_BUILDING_BASES",
+    "autolearn": false,
+    "never_learn": true,
+    "construction_blueprint": "fbmsw_brewery2_wood_southwest",
+    "blueprint_name": "southwest brewery2",
+    "blueprint_requires": [ { "id": "fbmsw_southwest" } ],
+    "blueprint_provides": [ { "id": "fbmsw_southwest" } ],
+    "blueprint_excludes": [ { "id": "fbmsw_souththwest" } ],
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_wood.json]: "blueprint_excludes": [ { "id": "fbmsw_souththwest" } ],

File: data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_wad.json (hunk 7 above)
Comment (Maleclypse) [reply data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_wad.json]: Ok I think I see what I needed to do there now.  I've made changes and will push. I've renamed the upgrades to fbms...1 and 2.

**Comment by curstwist** [comment]
You can also use "amount": ..either should work.

**Comment by Maleclypse** [comment]
I'm not sure what the following means
``src/json.cpp:111: Failed to visit member 'faction_owner' in JsonObject at line 2480:78509:
{
    "type": "mapgen",
    "om_terrain": "faction_base_saltworks_0",
    "method": "json",
    "faction_owner": { "id": "your_followers", "area": [ { "x": 0, "y": 0, "x2": 23, "y2": 23 } ] },
    "weight": 250,``

**Comment by curstwist** [comment]
Faction_owner needs to be within the object section. They were all wrong but jbtw recently fixed the others.

**Comment by AMurkin** [comment]
Shouldn't `faction_owner` be inside `object`?

File: data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_log.json [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    "result": "faction_base_modular_brewer2_log_southwest",
+    "description": "We could use fill out the space with extra kegs and storage racks.",
+    "category": "CC_BUILDING",
+    "subcategory": "CSC_BUILDING_BASES",
+    "autolearn": false,
+    "never_learn": true,
+    "construction_blueprint": "fbmsw_brewery2_log_southwest",
+    "blueprint_name": "southwest brewery2",
+    "blueprint_requires": [ { "id": "fbmsw_southwest1" } ],
+    "blueprint_provides": [ { "id": "fbmsw_southwest2" } ],
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_log.json]: excludes id is still misspelled. should be identical to requires.

File: data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_metal.json [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    "description": "We could use fill out the space with extra kegs and storage racks.",
+    "category": "CC_BUILDING",
+    "subcategory": "CSC_BUILDING_BASES",
+    "autolearn": false,
+    "never_learn": true,
+    "construction_blueprint": "fbmsw_brewery2_metal_southwest",
+    "blueprint_name": "southwest brewery2",
+    "blueprint_requires": [ { "id": "fbmsw_southwest1" } ],
+    "blueprint_provides": [ { "id": "fbmsw_southwest2" } ],
+    "blueprint_excludes": [ { "id": "fbmsw_souththwest2" } ],
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_metal.json]: excludes id is still misspelled. should be identical to requires.

File: data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_wad.json [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    "description": "We could use fill out the space with extra kegs and storage racks.",
+    "category": "CC_BUILDING",
+    "subcategory": "CSC_BUILDING_BASES",
+    "autolearn": false,
+    "never_learn": true,
+    "construction_blueprint": "fbmsw_brewery2_wad_southwest",
+    "blueprint_name": "southwest brewery2",
+    "blueprint_requires": [ { "id": "fbmsw_southwest1" } ],
+    "blueprint_provides": [ { "id": "fbmsw_southwest2" } ],
+    "blueprint_excludes": [ { "id": "fbmsw_souththwest2" } ],
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_wad.json]: fbmsw_souththwest2 should be fbmsw_southwest2

File: data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_stone.json [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    "description": "We could use fill out the space with extra kegs and storage racks.",
+    "category": "CC_BUILDING",
+    "subcategory": "CSC_BUILDING_BASES",
+    "autolearn": false,
+    "never_learn": true,
+    "construction_blueprint": "fbmsw_brewery2_rock_southwest",
+    "blueprint_name": "southwest brewery2",
+    "blueprint_requires": [ { "id": "fbmsw_southwest1" } ],
+    "blueprint_provides": [ { "id": "fbmsw_southwest2" } ],
+    "blueprint_excludes": [ { "id": "fbmsw_souththwest2" } ],
```
Comment (curstwist) [inline data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_stone.json]: excludes id is still misspelled. should be identical to requires.

File: data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_log.json (hunk 11 above)
Comment (Maleclypse) [reply data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_log.json]: I followed the same format from the evac shelter recipe? I think?
``    "blueprint_requires": [ { "id": "fbmc_shelter_bedroom1" } ],
    "blueprint_provides": [ { "id": "fbmc_shelter_bedroom2" } ],
    "blueprint_excludes": [ { "id": "fbmc_shelter_bedroom2" } ],``

Comment (curstwist) [reply data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_log.json]: Yeah that's fine, just check the spelling of southwest,  3 are souththwest

Comment (Maleclypse) [reply data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_log.json]: Ah! It's like that thing where as long as the first and last letter are right the brain just fills in the rest.  Thanks clarifying neither I or my wife could find the spelling error.

Comment (curstwist) [reply data/json/recipes/basecamps/recipe_modular_saltworks/recipe_modular_saltworks_log.json]: Yeah, it's hard to spot, the eye goes right past it : )

**Comment by Maleclypse** [comment]
I'm seeing a bunch of gunmod errors in my C and Travis tests but not anymore errors from files I edited.  Tomorrow I'll see if I can get the brewery upgrades to work and I'll post screenshots of them assuming they do.

**Comment by Maleclypse** [comment]
@curstwist this might just need another set of eyes or maybe I should take a break for the day on it but I've got all the material types building all the way through except for the final brewery upgrade on metal and wad types.  I've also solved my issue where it would offer me to change material types in the base construction screen, I'm not sure if it was the most elegant solution but it seems to have worked by editing the ``blueprint excludes`` line.

**Comment by Maleclypse** [comment]
Image of completed all materials versions here in Camp Salty.   Top left has a basic brewery, bottom right has upgraded brewery by one level.  All other breweries are level 2.  Left is log, top left is stone, north is metal, top right is wattle and daub, right is wood.  

Saltworks final

**Pull request merged**